portable/NetworkInterface/ThirdParty/MSP432/NetworkInterface.c

```diff
@@ -0,0 +1,981 @@
+                }
+            }
+        }
+
+        /* Release the DMA descriptor at a specific index */
+        taskENTER_CRITICAL();
+        DMAFreeDescriptorRX( NIDataReceived.ulindx );
+        taskEXIT_CRITICAL();
+    } /* end for */
+}
+
+
+/* Network initialization is already done outside of this function, so this function will only
+ * report if the network is initialized by returning a flag. Called directly by FreeRTOS. */
+BaseType_t xNetworkInterfaceInitialise( void )
+{
+    if( networkUP == true )
+    {
+        vTaskDelay( pdMS_TO_TICKS( ETH_UP_DELAY_MS ) );
+        return pdTRUE;
+    }
+
+    return pdFALSE;
+}
+
+
+/*  Task to output the data to the network interface. This allows the network stack to continue
+ *  processing while the data is able to be sent. */
+void prvEMACDeferredInterfaceOutput( void * pvParameters )
+{
+    struct NetworkInterfaceDataOut NIDataOutput;
+
+    for( ; ; )
+    {
+        ulTaskNotifyTakeIndexed( TX_QUEPOS_FIRST_EVENT, pdTRUE, portMAX_DELAY );
+        xQueueReceive( xQueueTX, &NIDataOutput, 0 );
+
+        /* For ICMP packets, the checksum must be zero if the network hardware computes the checksum or the buffer will not be properly sent
+         * For this driver it is required that ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM == 1 */
+        ProtocolPacket_t * pxPacket;
+        pxPacket = ( ProtocolPacket_t * ) ( NIDataOutput.pxDescriptor->pucEthernetBuffer );
+
+        if( pxPacket->xICMPPacket.xIPHeader.ucProtocol == ( uint8_t ) ipPROTOCOL_ICMP )
+        {
+            pxPacket->xICMPPacket.xICMPHeader.usChecksum = ( uint16_t ) 0u;
+        }
+
+        taskENTER_CRITICAL();
+        packetTransmit( NIDataOutput.pxDescriptor->pucEthernetBuffer, ( uint32_t ) NIDataOutput.pxDescriptor->xDataLength );
+        taskEXIT_CRITICAL();
+
+        /* wait until transmit */
+        ulTaskNotifyTakeIndexed( TX_QUEPOS_SECOND_EVENT, pdTRUE, portMAX_DELAY );
+        iptraceNETWORK_INTERFACE_TRANSMIT();
+
+        if( NIDataOutput.xReleaseAfterSend )
+        {
+            vReleaseNetworkBufferAndDescriptor( NIDataOutput.pxDescriptor );
+        }
+    }
+}
+
+
+/*  Function to output packets to the network.
+ *  Called directly by FreeRTOS. This function should not block and therefore passes
+ *  all output data to a queue.  If the function blocks the network stack can become unresponsive
+ *  so all of the output work is thereby done by another task. */
+BaseType_t xNetworkInterfaceOutput( NetworkBufferDescriptor_t * const pxDescriptor,
+                                    BaseType_t xReleaseAfterSend )
+{
+    struct NetworkInterfaceDataOut NIDataOutput;
+    BaseType_t txResp;
+
+    NIDataOutput.pxDescriptor = pxDescriptor;
+    NIDataOutput.xReleaseAfterSend = xReleaseAfterSend;
+    txResp = xQueueSend( xQueueTX, &NIDataOutput, 0 );
+
+    if( txResp == pdTRUE )
+    {
+        xTaskNotifyGiveIndexed( xTaskToNotifyEthernetTX, 0 );
+        return pdTRUE;
```

Comment:
nit: you can simplify this function to use a single `return` statement by introducing a helper variable, e.g.:

```c
BaseType_t returnValue = pdFALSE;
.
.
.
if( txResp == pdTRUE )
{
    xTasknotifyGiveIndexed( xTaskToNotifyEthernetTX, 0 );
    returnValue = pdTRUE;
}
return returnValue;
```

Reply:
Yes, this is a good idea.  The code is updated.